### PR TITLE
fix RecordWriter Flush only work when ZlibCompressed

### PR DESCRIPTION
**Pull request by fo40225**
Didn't know why RecordWriter.Flush() calls dest_.Flush() only when compressed.

WritableFile has Flush() function.

related: 
https://github.com/tensorflow/tensorflow/issues/26683

### CODE REVIEW

**Review by alextp — COMMENTED** [review]
Wow I can't believe we had this bug! Good catch!

Can you add a unit test so we don't accidentally have this issue again?

**Comment by fo40225** [comment]
It already have `TEST(RecordReaderWriterTest, TestFlush) {` in `record_reader_writer_test.cc`

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/lib/io/record_reader_writer_test.cc

**Comment by alextp** [comment]
Yes but clearly this test did not test that flushing was happening correctly when flush was called, as per the bug you found.

What I'd like to see is a new test in that file that fails before your change but passes after your change. This will guarantee that future developers don't accidentally undo the work you just did.

**Review by alextp — COMMENTED** [review]
Does this test fail without this PR?

**Comment by fo40225** [comment]
After test, not only uncompressed file flush failed but also compressed file flush failed.

Need more fix lol.

v1.13.1 + unit test
```
user@ubuntu:~/tensorflow$ bazel test --config=opt --test_output=errors //tensorflow/python:tf_record_test
INFO: Invocation ID: 4ce9511e-5fb7-4e8f-9868-cf5ed185856a
WARNING: /home/user/tensorflow/tensorflow/python/BUILD:2986:1: in py_library rule //tensorflow/python:standard_ops: target '//tensorflow/python:standard_ops' depends on deprecated target '//tensorflow/python/ops/distributions:distributions': TensorFlow Distributions has migrated to TensorFlow Probability (https://github.com/tensorflow/probability). Deprecated copies remaining in tf.distributions will not receive new features, and will be removed by early 2019. You should update all usage of `tf.distributions` to `tfp.distributions`.
INFO: Analysed target //tensorflow/python:tf_record_test (0 packages loaded, 0 targets configured).
INFO: Found 1 test target...
FAIL: //tensorflow/python:tf_record_test (see /home/user/.cache/bazel/_bazel_user/f9f5c94ad789d21bcfdb930889866229/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/tf_record_test/test.log)
INFO: From Testing //tensorflow/python:tf_record_test:
==================== Test output for //tensorflow/python:tf_record_test:
sWARNING:tensorflow:From /home/user/.cache/bazel/_bazel_user/f9f5c94ad789d21bcfdb930889866229/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/tf_record_test.runfiles/org_tensorflow/tensorflow/python/lib/io/tf_record_test.py:419: tf_record_iterator (from tensorflow.python.lib.io.tf_record) is deprecated and will be removed in a future version.
Instructions for updating:
Use eager execution and:
`tf.data.TFRecordDataset(path)`
.....s.F....s.F....s.F....s.....s....s
======================================================================
FAIL: testFlush (__main__.TFRecordWriterCloseAndFlushGzipTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/user/.cache/bazel/_bazel_user/f9f5c94ad789d21bcfdb930889866229/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/tf_record_test.runfiles/org_tensorflow/tensorflow/python/lib/io/tf_record_test.py", line 461, in testFlush
    self.assertListEqual(actual, records)
AssertionError: Lists differ: [] != ['Record 0', 'Record 1', 'Reco...

Second list contains 20 additional elements.
First extra element 0:
'Record 0'

- []
+ ['Record 0',
+  'Record 1',
+  'Record 2',
+  'Record 3',
+  'Record 4',
+  'Record 5',
+  'Record 6',
+  'Record 7',
+  'Record 8',
+  'Record 9',
+  'Record 10',
+  'Record 11',
+  'Record 12',
+  'Record 13',
+  'Record 14',
+  'Record 15',
+  'Record 16',
+  'Record 17',
+  'Record 18',
+  'Record 19']

======================================================================
FAIL: testFlush (__main__.TFRecordWriterCloseAndFlushTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/user/.cache/bazel/_bazel_user/f9f5c94ad789d21bcfdb930889866229/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/tf_record_test.runfiles/org_tensorflow/tensorflow/python/lib/io/tf_record_test.py", line 461, in testFlush
    self.assertListEqual(actual, records)
AssertionError: Lists differ: [] != ['Record 0', 'Record 1', 'Reco...

Second list contains 20 additional elements.
First extra element 0:
'Record 0'

- []
+ ['Record 0',
+  'Record 1',
+  'Record 2',
+  'Record 3',
+  'Record 4',
+  'Record 5',
+  'Record 6',
+  'Record 7',
+  'Record 8',
+  'Record 9',
+  'Record 10',
+  'Record 11',
+  'Record 12',
+  'Record 13',
+  'Record 14',
+  'Record 15',
+  'Record 16',
+  'Record 17',
+  'Record 18',
+  'Record 19']

======================================================================
FAIL: testFlush (__main__.TFRecordWriterCloseAndFlushZlibTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/user/.cache/bazel/_bazel_user/f9f5c94ad789d21bcfdb930889866229/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/tf_record_test.runfiles/org_tensorflow/tensorflow/python/lib/io/tf_record_test.py", line 461, in testFlush
    self.assertListEqual(actual, records)
AssertionError: Lists differ: [] != ['Record 0', 'Record 1', 'Reco...

Second list contains 20 additional elements.
First extra element 0:
'Record 0'

- []
+ ['Record 0',
+  'Record 1',
+  'Record 2',
+  'Record 3',
+  'Record 4',
+  'Record 5',
+  'Record 6',
+  'Record 7',
+  'Record 8',
+  'Record 9',
+  'Record 10',
+  'Record 11',
+  'Record 12',
+  'Record 13',
+  'Record 14',
+  'Record 15',
+  'Record 16',
+  'Record 17',
+  'Record 18',
+  'Record 19']

----------------------------------------------------------------------
Ran 39 tests in 0.920s

FAILED (failures=3, skipped=7)
================================================================================
Target //tensorflow/python:tf_record_test up-to-date:
  bazel-bin/tensorflow/python/tf_record_test
INFO: Elapsed time: 2.396s, Critical Path: 1.79s
INFO: 1 process: 1 local.
INFO: Build completed, 1 test FAILED, 2 total actions
//tensorflow/python:tf_record_test                                       FAILED in 1.8s
  /home/user/.cache/bazel/_bazel_user/f9f5c94ad789d21bcfdb930889866229/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/tf_record_test/test.log

INFO: Build completed, 1 test FAILED, 2 total actions
```

v1.13.1 + db960e9
```
user@TR4:~/tensorflow$ bazel test --config=opt --test_output=errors //tensorflow/python:tf_record_test
INFO: Invocation ID: 6ee89cb4-377a-4c9d-b47a-b8d14dfd1789
WARNING: /home/user/tensorflow/tensorflow/python/BUILD:2986:1: in py_library rule //tensorflow/python:standard_ops: target '//tensorflow/python:standard_ops' depends on deprecated target '//tensorflow/python/ops/distributions:distributions': TensorFlow Distributions has migrated to TensorFlow Probability (https://github.com/tensorflow/probability). Deprecated copies remaining in tf.distributions will not receive new features, and will be removed by early 2019. You should update all usage of `tf.distributions` to `tfp.distributions`.
INFO: Analysed target //tensorflow/python:tf_record_test (0 packages loaded, 0 targets configured).
INFO: Found 1 test target...
FAIL: //tensorflow/python:tf_record_test (see /home/user/.cache/bazel/_bazel_user/f9f5c94ad789d21bcfdb930889866229/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/tf_record_test/test.log)
INFO: From Testing //tensorflow/python:tf_record_test:
==================== Test output for //tensorflow/python:tf_record_test:
sWARNING:tensorflow:From /home/user/.cache/bazel/_bazel_user/f9f5c94ad789d21bcfdb930889866229/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/tf_record_test.runfiles/org_tensorflow/tensorflow/python/lib/io/tf_record_test.py:419: tf_record_iterator (from tensorflow.python.lib.io.tf_record) is deprecated and will be removed in a future version.
Instructions for updating:
Use eager execution and:
`tf.data.TFRecordDataset(path)`
.....s.F....s......s.F....s.....s....s
======================================================================
FAIL: testFlush (__main__.TFRecordWriterCloseAndFlushGzipTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/user/.cache/bazel/_bazel_user/f9f5c94ad789d21bcfdb930889866229/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/tf_record_test.runfiles/org_tensorflow/tensorflow/python/lib/io/tf_record_test.py", line 461, in testFlush
    self.assertListEqual(actual, records)
AssertionError: Lists differ: [] != ['Record 0', 'Record 1', 'Reco...

Second list contains 20 additional elements.
First extra element 0:
'Record 0'

- []
+ ['Record 0',
+  'Record 1',
+  'Record 2',
+  'Record 3',
+  'Record 4',
+  'Record 5',
+  'Record 6',
+  'Record 7',
+  'Record 8',
+  'Record 9',
+  'Record 10',
+  'Record 11',
+  'Record 12',
+  'Record 13',
+  'Record 14',
+  'Record 15',
+  'Record 16',
+  'Record 17',
+  'Record 18',
+  'Record 19']

======================================================================
FAIL: testFlush (__main__.TFRecordWriterCloseAndFlushZlibTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/user/.cache/bazel/_bazel_user/f9f5c94ad789d21bcfdb930889866229/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/tf_record_test.runfiles/org_tensorflow/tensorflow/python/lib/io/tf_record_test.py", line 461, in testFlush
    self.assertListEqual(actual, records)
AssertionError: Lists differ: [] != ['Record 0', 'Record 1', 'Reco...

Second list contains 20 additional elements.
First extra element 0:
'Record 0'

- []
+ ['Record 0',
+  'Record 1',
+  'Record 2',
+  'Record 3',
+  'Record 4',
+  'Record 5',
+  'Record 6',
+  'Record 7',
+  'Record 8',
+  'Record 9',
+  'Record 10',
+  'Record 11',
+  'Record 12',
+  'Record 13',
+  'Record 14',
+  'Record 15',
+  'Record 16',
+  'Record 17',
+  'Record 18',
+  'Record 19']

----------------------------------------------------------------------
Ran 39 tests in 0.904s

FAILED (failures=2, skipped=7)
================================================================================
Target //tensorflow/python:tf_record_test up-to-date:
  bazel-bin/tensorflow/python/tf_record_test
INFO: Elapsed time: 2.396s, Critical Path: 1.74s
INFO: 1 process: 1 local.
INFO: Build completed, 1 test FAILED, 2 total actions
//tensorflow/python:tf_record_test                                       FAILED in 1.7s
  /home/user/.cache/bazel/_bazel_user/f9f5c94ad789d21bcfdb930889866229/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/tf_record_test/test.log

INFO: Build completed, 1 test FAILED, 2 total actions
```

**Review by alextp — DISMISSED** [review]

**Comment by alextp** [comment]
Yes, this PR does not currently look ready to merge. Please update when it is ready for another look.

**Comment by rthadur** [comment]
> Yes, this PR does not currently look ready to merge. Please update when it is ready for another look.

@alextp can we merge this PR ?

**Comment by alextp** [comment]
@rthadur no, there have been no updates since I last looked at it

**Review by alextp — DISMISSED** [review]

**Comment by rthadur** [comment]
@fo40225 can you please check build failures

**Comment by rthadur** [comment]
@alextp can you please review new changes

**Review by alextp — DISMISSED** [review]
Approving to run tests

File: tensorflow/python/lib/io/tf_record_test.py [hunk 1]
```diff
@@ -447,6 +448,18 @@ def testWriteAndRead(self):
     actual = list(tf_record.tf_record_iterator(self._fn, self._options))
     self.assertListEqual(actual, records)
 
+  def testFlush(self):
+    records = list(map(self._Record, range(self._num_records)))
+    def childProcess(writer, rs):
```
Comment (rthadur) [inline tensorflow/python/lib/io/tf_record_test.py]: Function name "childProcess" doesn't conform to u'^(?:(?P<exempt>setUp|tearDown|setUpModule|tearDownModule)|(?P<camel_case>_?[A-Z][a-zA-Z0-9]*)|(?P<snake_case>_?[a-z][a-z0-9_]*))$' pattern [invalid-name]

**Review by rthadur — CHANGES_REQUESTED** [review]

File: tensorflow/python/lib/io/tf_record_test.py (hunk 1 above)
Comment (fo40225) [reply tensorflow/python/lib/io/tf_record_test.py]: Does this mean that I should change `childProcess` to `ChildProcess` or `child_process`?

Comment (alextp) [reply tensorflow/python/lib/io/tf_record_test.py]: Just add a # pylint: disable=invalid-name

**Review by alextp — DISMISSED** [review]

**Comment by rthadur** [comment]
@fo40225 can you please check build failures ?

**Comment by fo40225** [comment]
- Ubuntu Sanity
I have fixed the pylint error.

- Ubuntu CC
I have no idea about this timeout, please provide the detail error log.
```
TIMEOUT: //tensorflow/core:lib_io_record_reader_writer_test (Summary)
      /tmpfs/tmp/bazel/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/core/lib_io_record_reader_writer_test/test.log
      /tmpfs/tmp/bazel/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/core/lib_io_record_reader_writer_test/test.log
      /tmpfs/tmp/bazel/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/core/lib_io_record_reader_writer_test/test_attempts/attempt_1.log
      /tmpfs/tmp/bazel/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/core/lib_io_record_reader_writer_test/test.log
      /tmpfs/tmp/bazel/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/core/lib_io_record_reader_writer_test/test_attempts/attempt_2.log
```

---

Those error seems not related to this PR:
- Linux GPU
`[  FAILED  ] IndexedCaseTest.testParallelExecution`

- MacOS Contrib
`tensorflow/lite/delegates/nnapi/nnapi_delegate.cc:764:36: error: use of undeclared identifier 'PROT_READ'`

- Windows Bazel & Windows Bazel GPU
`tensorflow/lite/mmap_allocation_disabled.cc(24): error C2065: 'kMMap': undeclared identifier`

**Review by alextp — DISMISSED** [review]

**Review by alextp — APPROVED** [review]

**Pull request merged**